### PR TITLE
be a nice client

### PR DESCRIPTION
**Pull request by tburgin**
*  Add configuration for max startup jitter and backoff
*  Add test_client binary example

### CODE REVIEW

File: MOLFCMClient/MOLFCMClient.m [hunk 1]
```diff
@@ -82,19 +91,23 @@ @implementation MOLFCMClient
...
+                 connectDelayMax:(uint32_t)connectDelayMax
+                      backoffMax:(uint32_t)backoffMax
+                      fatalCodes:(NSArray<NSNumber *> *)fatalCodes
             sessionConfiguration:(NSURLSessionConfiguration *)sessionConfiguration
                   messageHandler:(MOLFCMMessageHandler)messageHandler {
   self = [super init];
   if (self) {
     _FCMToken = [FCMToken copy];
-    _bindComponents = [NSURLComponents componentsWithString:kFCMHost];
+    _bindComponents = [NSURLComponents componentsWithString:host ?: kFCMHost];
```
Comment (russellhancox) [inline MOLFCMClient/MOLFCMClient.m]: host is not marked nonnull, passing nil will throw an exception.

File: MOLFCMClient/MOLFCMClient.h [hunk 2]
```diff
@@ -42,23 +48,39 @@ typedef void (^MOLFCMAcknowledgeErrorHandler)(NSDictionary *, NSError *);
...
  */
 - (instancetype)initWithFCMToken:(NSString *)FCMToken
+                            host:(NSString *)host
+                 connectDelayMax:(uint32_t)connectDelayMax
+                      backoffMax:(uint32_t)backoffMax
+                      fatalCodes:(NSArray<NSNumber *> *)fatalCodes
             sessionConfiguration:(NSURLSessionConfiguration *)sessionConfiguration
-                  messageHandler:(MOLFCMMessageHandler)messageHandler
-                     NS_DESIGNATED_INITIALIZER;
+                  messageHandler:(MOLFCMMessageHandler)messageHandler NS_DESIGNATED_INITIALIZER;
```
Comment (russellhancox) [inline MOLFCMClient/MOLFCMClient.h]: If possible the initializer you're replacing should still exist and just call the new one

File: MOLFCMClient/MOLFCMClient.m [hunk 3]
```diff
@@ -31,9 +31,12 @@
 static NSString *const kFCMApplicationJSON = @"application/json";
 static NSString *const kFCMContentType = @"Content-Type";
 
-/**  15 minute backoff maximum */
+/**  Default 15 minute backoff maximum */
 static const uint32_t kBackoffMaxSeconds = 900;
 
+/**  Default 10 sec connect delay maximum */
+static const uint32_t kConnectDelayMaxSeconds = 10;
```
Comment (russellhancox) [inline MOLFCMClient/MOLFCMClient.m]: These 2 consts would be better having default in the name

File: MOLFCMClient/MOLFCMClient.m [hunk 4]
```diff
@@ -82,19 +91,23 @@ @implementation MOLFCMClient
 #pragma mark init/dealloc methods
 
 - (instancetype)initWithFCMToken:(NSString *)FCMToken
+                            host:(NSString *)host
+                 connectDelayMax:(uint32_t)connectDelayMax
+                      backoffMax:(uint32_t)backoffMax
+                      fatalCodes:(NSArray<NSNumber *> *)fatalCodes
```
Comment (russellhancox) [inline MOLFCMClient/MOLFCMClient.m]: qq: why is this overrideable? presumably as a client of FCM this class should know best about status codes.

File: MOLFCMClient/MOLFCMClient.m (hunk 1 above)
Comment (tburgin) [reply MOLFCMClient/MOLFCMClient.m]: Don't think so elvis.

File: MOLFCMClient/MOLFCMClient.m (hunk 4 above)
Comment (tburgin) [reply MOLFCMClient/MOLFCMClient.m]: You would think. This is mainly so we can change them without having to do a release here.

File: MOLFCMClient/MOLFCMClient.h (hunk 2 above)
Comment (tburgin) [reply MOLFCMClient/MOLFCMClient.h]: sure thing

File: MOLFCMClient/MOLFCMClient.m (hunk 3 above)
Comment (tburgin) [reply MOLFCMClient/MOLFCMClient.m]: Good idea

File: MOLFCMClient/MOLFCMClient.m (hunk 1 above)
Comment (russellhancox) [reply MOLFCMClient/MOLFCMClient.m]: Oh oops.

**Pull request merged**